### PR TITLE
Raven db 12277

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Server/Documents/Queries/Graph/EdgeQueryStep.cs [hunk 1]
```diff
@@ -65,9 +65,16 @@ public bool IsEmpty()
             return _results.Count == 0;
         }
 
+        public bool CollectIntermediateResults { get; set; }
+
+        public List<Match> IntermediateResults => CollectIntermediateResults?_results: new List<GraphQueryRunner.Match>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Graph/EdgeQueryStep.cs:70]: I don't like this. It just allocates and throw it away.
Why do we need this to be valid at all if `CollectIntermediateResults` is `false`?

File: src/Raven.Server/Documents/Queries/Graph/RecursionQueryStep.cs [hunk 2]
```diff
@@ -109,9 +109,16 @@ public bool IsEmpty()
             return _results.Count == 0;
         }
 
+        public bool CollectIntermediateResults { get; set; }
+
+        public List<Match> IntermediateResults => CollectIntermediateResults ? _results : new List<Match>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Graph/RecursionQueryStep.cs:114]: In fact, why do we need it to be different from `_results` ?

File: src/Raven.Server/Documents/Queries/Graph/EdgeQueryStep.cs (hunk 1 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/Queries/Graph/EdgeQueryStep.cs:70]: we could return null but I would still want an indication that the intermediate results are needed, see: https://issues.hibernatingrhinos.com/issue/RavenDB-12572

File: src/Raven.Server/Documents/Queries/Graph/RecursionQueryStep.cs (hunk 2 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/Queries/Graph/RecursionQueryStep.cs:114]: collection destination doesn't have _result and we need to collect the requested documents.

**Pull request merged**